### DISCUSSION TITLE
Get routes from cms (api)?

### DISCUSSION THREAD

**Discussion by warloque** [react-static/react-static#1439] [General] [answered]
Any thoughts are welcome on how to get routes dynamically instead of listing them manually in static.config.js. 
Is it even possible with react-static or it goues against its nature?

**Comment by nddery** [comment, marked as answer]
The [`getRoutes`](https://github.com/react-static/react-static/blob/master/docs/config.md#getroutes) function is `async`, so you could fetch your routes from your CMS, then loop over them to create [`route`](https://github.com/react-static/react-static/blob/master/docs/config.md#route) objects. This is what's done in the [basic template](https://github.com/react-static/react-static/blob/master/packages/react-static/templates/basic/static.config.js) for the individual blog posts.

```js
// static.config.js
export default {
  getRoutes: async () => {
    const cmsRoutes = await fetchCmsRoutes()
    return [
      ...yourNormalStaticRoutes,
      ...cmsRoutes.map(route => ({
        path: route.slug,
        template: `src/templates/${route.template}.js`,
        ...
      })
    ]
  }
}
```

One thing to be aware of is that during development (`yarn start`), `fetchCmsRoutes()` will be called before the development server can start. So you'll want to be mindful what's retrieved outside of individual routes `getData` function.

**Reply by warloque** [reply]
Thanks for your snippet! This could be it, needs to be tested with dynamic routes though. 
Another part of routes syncing task is now running `build` upon `fetchCmsRoutes` updates... but that goes into deploy setup, I suppose.

**Reply by nddery** [reply]
When you mention dynamic routes, do you mean routes that aren't static, ie.: runtime only routes - like [these](https://github.com/react-static/react-static/blob/master/docs/guides/dynamic-routes-reach-router.md) ?

**Reply by warloque** [reply]
Yes, exactly, I meant react-static 'dynamic' routes.

**Reply by nddery** [reply]
Ah I see, I had misunderstood your question originally then, sorry about that. I feel like this would go against the idea of a static site generator.

If these paths aren't known upfront, maybe you could validate them against your API inside your 404 template ? 

```jsx
const TEMPLATES = {
  '404': React.lazy(() => import('./NotFound')),
  'BlogPost': React.lazy(() => import('./BlogPost'))
}

export default function NotFound() { // This is the fallback, when no static route has matched
  const [route, setRoute] = React.useState(false)

  React.useEffect(() => {
    fetchRouteFromAPI(window.location.href)
      .then(setRoute)
  }, [])

  if (!route) return null

  const Template = TEMPLATES[route.template]
  return <Template data={route.data} />
}
```

This will lead to a waterfall of requests, so probably not the best for performance.

**Reply by SleeplessByte** [reply]
Yeah, the idea here would be to have a webhook in your CMS trigger a _build_ so you have a static site again. You can definitely do it with dynamic routes, but regular react might be better then.

**Reply by warloque** [reply]
Thanks, guys, for your input!
That's exactly what I'm trying to do – build static RS-app using API of a headless CMS. 
I came to the conclusion that CRA suits my task better at this point, so I leave RS for the landing app serving more or less static content and setup basic react app for the main app.
